### ISSUE TITLE
Warnings for `in`/`readonly ref`

### ISSUE THREAD

**Issue by benaadams** [dotnet/roslyn#23332]
Primitive value-types that will be passed by register (`byte`, `short`, `int`, etc) should generate a warning when passed as `in` as passing by readonly reference in the common case is likely a deoptimization error (dereference vs register)

Reference-types should generate a warning when passed as `in` as passing by readonly reference in the common case is likely a deoptimization error (double dereference vs dereference)

/cc @mikedn

**Comment by jcouv** [comment]
Tagging @VSadov

**Comment by VSadov** [comment]
There are cases where having an `in` parameter typed as reference type makes sense. 
Like ```Volatile.Read(in object variable)```

In general it is indeed not very useful, but there would be a need for a whitelist of few reasonable uses. 
I could be a good idea for an analyzer.

**Comment by benaadams** [comment]
> There are cases where having an in parameter typed as reference type makes sense.
> Like `Volatile.Read(in object variable)`

But they could be #pragma'd as valid?

**Comment by tannergooding** [comment]
I agree this would be a good case for an analyzer.

That being said, for some types whether or not it should be passed as 'in' depends on the underlying architecture and calling convention. So the diagnostics should likely only cover the well-defined cases that are consistent between all targets.

**Comment by MaStr11** [comment]
This should be considered for the warning waves #1580.
